In [2]:
import pandas as pd

In [4]:
news = pd.read_csv('raw_analyst_ratings.csv')

In [5]:
news

Unnamed: 0                                           headline  \
0                 0            Stocks That Hit 52-Week Highs On Friday   
1                 1         Stocks That Hit 52-Week Highs On Wednesday   
2                 2                      71 Biggest Movers From Friday   
3                 3       46 Stocks Moving In Friday's Mid-Day Session   
4                 4  B of A Securities Maintains Neutral on Agilent...   
...             ...                                                ...   
1407323     1413844             Top Narrow Based Indexes For August 29   
1407324     1413845  Recap: Wednesday's Top Percentage Gainers and ...   
1407325     1413846  UPDATE: Oppenheimer Color on China Zenix Auto ...   
1407326     1413847  Oppenheimer Initiates China Zenix At Outperfor...   
1407327     1413848  China Zenix Auto International Opens For Tradi...   

                                                       url          publisher  \
0        https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1        https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2        https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3        https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4        https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   
...                                                    ...                ...   
1407323  https://www.benzinga.com/news/11/08/1888782/to...      Monica Gerson   
1407324  https://www.benzinga.com/news/earnings/11/06/1...       Benjamin Lee   
1407325  https://www.benzinga.com/analyst-ratings/analy...     BenzingaStaffL   
1407326  https://www.benzinga.com/analyst-ratings/price...          Joe Young   
1407327  https://www.benzinga.com/news/ipos/11/05/10789...      Allie Wickman   

                              date stock  
0        2020-06-05 10:30:54-04:00     A  
1        2020-06-03 10:45:20-04:00     A  
2        2020-05-26 04:30:07-04:00     A  
3        2020-05-22 12:45:06-04:00     A  
4        2020-05-22 11:38:59-04:00     A  
...                            ...   ...  
1407323        2011-08-29 00:00:00    ZX  
1407324        2011-06-22 00:00:00    ZX  
1407325        2011-06-21 00:00:00    ZX  
1407326        2011-06-21 00:00:00    ZX  
1407327        2011-05-12 00:00:00    ZX  

[1407328 rows x 6 columns]

In [10]:
news.head()

Unnamed: 0                                           headline  \
0           0            Stocks That Hit 52-Week Highs On Friday   
1           1         Stocks That Hit 52-Week Highs On Wednesday   
2           2                      71 Biggest Movers From Friday   
3           3       46 Stocks Moving In Friday's Mid-Day Session   
4           4  B of A Securities Maintains Neutral on Agilent...   

                                                 url          publisher  \
0  https://www.benzinga.com/news/20/06/16190091/s...  Benzinga Insights   
1  https://www.benzinga.com/news/20/06/16170189/s...  Benzinga Insights   
2  https://www.benzinga.com/news/20/05/16103463/7...         Lisa Levin   
3  https://www.benzinga.com/news/20/05/16095921/4...         Lisa Levin   
4  https://www.benzinga.com/news/20/05/16095304/b...         Vick Meyer   

                        date stock  
0  2020-06-05 10:30:54-04:00     A  
1  2020-06-03 10:45:20-04:00     A  
2  2020-05-26 04:30:07-04:00     A  
3  2020-05-22 12:45:06-04:00     A  
4  2020-05-22 11:38:59-04:00     A

In [7]:
news.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1407328 entries, 0 to 1407327
Data columns (total 6 columns):
 #   Column      Non-Null Count    Dtype 
---  ------      --------------    ----- 
 0   Unnamed: 0  1407328 non-null  int64 
 1   headline    1407328 non-null  object
 2   url         1407328 non-null  object
 3   publisher   1407328 non-null  object
 4   date        1407328 non-null  object
 5   stock       1407328 non-null  object
dtypes: int64(1), object(5)
memory usage: 64.4+ MB


In [11]:
 missing_values = news.isnull().sum()
missing_values[missing_values > 0]



Series([], dtype: int64)

So, from the above we can see the basic structure of the data provided.<br>This data has total number of 1407328 rows and 6 column<br>all the useful data types are objects<br>There is no messing value provided in the data so I can say that this is a clean data

In [16]:
news.columns

Index(['Unnamed: 0', 'headline', 'url', 'publisher', 'date', 'stock'], dtype='object')

# Descriptive Analysis<br>
On the descriptive analysis I will go through the understanding of of each columns information

On the head line column first I will go through text data analysis

In [17]:
unique_headlines = news['headline'].nunique()
print(f"Number of unique headlines: {unique_headlines}")


Number of unique headlines: 845770


In [18]:
news['headline_length'] = news['headline'].apply(len)
avg_headline_length = news['headline_length'].mean()
print(f"Average length of headlines: {avg_headline_length:.2f} characters")



Average length of headlines: 73.12 characters


In [20]:
common_words = news['headline'].str.lower().str.contains('stocks|market').sum()
print(f"Number of headlines containing 'stocks' or 'market': {common_words}")


Number of headlines containing 'stocks' or 'market': 280592


In [22]:

unique_urls = news['url'].nunique()
print(f"Number of unique URLs: {unique_urls}")


Number of unique URLs: 883429


In [23]:

news['domain'] = news['url'].apply(lambda x: x.split('/')[2])
domain_counts = news['domain'].value_counts()
print("Top 5 most frequent domains:")
print(domain_counts.head())

Top 5 most frequent domains:
domain
www.benzinga.com    1407328
Name: count, dtype: int64


In [24]:
unique_publishers = news['publisher'].nunique()
print(f"Number of unique publishers: {unique_publishers}")


Number of unique publishers: 1034


In [26]:
publisher_counts = news['publisher'].value_counts()
print("Top 5 most frequent publishers:")
print(publisher_counts.head())

Top 5 most frequent publishers:
publisher
Paul Quintaro        228373
Lisa Levin           186979
Benzinga Newsdesk    150484
Charles Gross         96732
Monica Gerson         82380
Name: count, dtype: int64


The Date entries need to convert to Date time format because it is given in string object type

In [53]:
news['date'] = pd.to_datetime(news['date'], errors='coerce', utc='True')

In [54]:
date_range = news['date'].agg([min, max])
print(f"Date range: {date_range['min']} to {date_range['max']}")

C:\Users\mama\AppData\Local\Temp\ipykernel_2368\1649777631.py:1: FutureWarning: The provided callable <built-in function min> is currently using Series.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  date_range = news['date'].agg([min, max])
C:\Users\mama\AppData\Local\Temp\ipykernel_2368\1649777631.py:1: FutureWarning: The provided callable <built-in function max> is currently using Series.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  date_range = news['date'].agg([min, max])


Date range: 2011-04-28 01:01:48+00:00 to 2020-06-11 21:12:35+00:00


In [56]:
# Frequency of articles by month/year
articles_per_month = news['date'].dt.to_period('M').value_counts().sort_index()
print("Number of articles per month:")
print(articles_per_month)


C:\Users\mama\AppData\Local\Temp\ipykernel_2368\1108811487.py:2: UserWarning: Converting to PeriodArray/Index representation will drop timezone information.
  articles_per_month = news['date'].dt.to_period('M').value_counts().sort_index()


Number of articles per month:
date
2011-04       6
2011-05      78
2011-06      95
2011-07     137
2011-08     149
           ... 
2020-02    2355
2020-03    4681
2020-04    4616
2020-05    9333
2020-06    5990
Freq: M, Name: count, Length: 111, dtype: int64


From the Data Stock colomn lets get information

In [57]:
unique_stocks = news['stock'].nunique()
print(f"Number of unique stocks mentioned: {unique_stocks}")


Number of unique stocks mentioned: 6204


In [58]:
stock_counts = news['stock'].value_counts()
print("Top 5 most frequently mentioned stocks:")
print(stock_counts.head())

Top 5 most frequently mentioned stocks:
stock
MRK     3333
MS      3238
NVDA    3146
MU      3142
QQQ     3106
Name: count, dtype: int64


In [ ]:
from itertools import combinations
from collections import Counter